src/mixins/itext_behavior.mixin.js

```diff
@@ -417,6 +424,270 @@
+        ctx.scale(1 / retinaScaling, 1 / retinaScaling);
+        ctx.drawImage(dragImage, 0, 0);
+        dragImage = c;
+      }
+      this.__dragImageDisposer && this.__dragImageDisposer();
+      this.__dragImageDisposer = function () {
+        dragImage.remove();
+      };
+      //  position drag image offsecreen
+      fabric.util.setStyle(dragImage, {
+        position: 'absolute',
+        left: -dragImage.width + 'px',
+        border: 'none'
+      });
+      fabric.document.body.appendChild(dragImage);
+      e.dataTransfer.setDragImage(dragImage, offset.x, offset.y);
+    },
+
+    /**
+     * support native like text dragging
+     * @private
+     * @param {DragEvent} e
+     * @returns {boolean} should handle event
+     */
+    onDragStart: function (e) {
+      this.__dragStartFired = true;
+      if (this.__isDragging) {
+        var selection = this.__dragStartSelection = {
+          selectionStart: this.selectionStart,
+          selectionEnd: this.selectionEnd,
+        };
+        var value = this._text.slice(selection.selectionStart, selection.selectionEnd).join('');
+        var data = Object.assign({ text: this.text, value: value }, selection);
+        e.dataTransfer.setData('text/plain', value);
+        e.dataTransfer.setData('application/fabric', JSON.stringify({
+          value: value,
+          styles: this.getSelectionStyles(selection.selectionStart, selection.selectionEnd, true)
+        }));
+        e.dataTransfer.effectAllowed = 'copyMove';
+        this.setDragImage(e, data);
+      }
+      return this.__isDragging;
+    },
+
+    /**
+     * Override to customize drag and drop behavior
+     * @public
+     * @param {DragEvent} e
+     * @returns {boolean}
+     */
+    canDrop: function (e) {
+      if (this.editable && !this.__corner) {
+        if (this.__isDragging && this.__dragStartSelection) {
+          //  drag source trying to drop over itself
+          //  allow dropping only outside of drag start selection
+          var index = this.getSelectionStartFromPointer(e);
+          var dragStartSelection = this.__dragStartSelection;
+          return index < dragStartSelection.selectionStart || index > dragStartSelection.selectionEnd;
+        }
+        return true;
+      }
+      return false;
+    },
+
+    /**
+     * support native like text dragging
+     * @private
+     * @param {object} options
+     * @param {DragEvent} options.e
+     */
+    dragOverHandler: function (options) {
+      var e = options.e;
+      var canDrop = !e.defaultPrevented && this.canDrop(e);
+      if (!this.__isDraggingOver && canDrop) {
+        this.__isDraggingOver = true;
+        this.enterEditing(e);
+        this.__isDragging && this.abortCursorAnimation();
+      }
+      else if (this.__isDraggingOver && !canDrop) {
+        //  drop state has changed
+        this.__isDraggingOver = false;
+        !this.__isDragging && this.clearContextTop();
+        this.exitEditing();
+      }
```

Comment:
couldn't this go on a dragEnter event? Wouldn't it be its natural place?

Reply:
I don't remember why I use the drag over event.
Perhaps it was before I fixed enter/leave firing order, so at the time it was unreliable.

Reply:
I think I understand now... 
Consider the edge case where an object is actively dragged over and `canDrop` is toggled during the drag over.
I think the object remains the target of the event so this case won't be handled from the `dragenter` event.

Reply:
I decided to add but not remove
I think is best this way
7777841e